django_mongodb_backend/indexes.py

```diff
@@ -101,6 +105,169 @@ def where_node_idx(self, compiler, connection):
+            return None
+        fields = {}
+        for field_name, _ in self.fields_orders:
+            field_ = model._meta.get_field(field_name)
+            type_ = self.search_index_data_types(field_, field_.db_type(schema_editor.connection))
+            field_path = column_prefix + model._meta.get_field(field_name).column
+            fields[field_path] = {"type": type_}
+        return SearchIndexModel(
+            definition={"mappings": {"dynamic": False, "fields": fields}}, name=self.name
+        )
+
+
+class VectorSearchIndex(SearchIndex):
+    suffix = "vsi"
+    ALLOWED_SIMILARITY_FUNCTIONS = frozenset(("euclidean", "cosine", "dotProduct"))
+    _error_id_prefix = "django_mongodb_backend.indexes.VectorSearchIndex"
+
+    def __init__(self, *expressions, similarities="cosine", **kwargs):
+        super().__init__(*expressions, **kwargs)
+        self.similarities = similarities
+
+    def check(self, model, connection):
+        errors = super().check(model, connection)
+        similarities = (
+            self.similarities if isinstance(self.similarities, list) else [self.similarities]
+        )
+        for func in similarities:
+            if func not in self.ALLOWED_SIMILARITY_FUNCTIONS:
+                errors.append(
+                    Error(
+                        f"{func} isn't a valid similarity function, options "
+                        f"are {', '.join(sorted(self.ALLOWED_SIMILARITY_FUNCTIONS))}",
+                        obj=self,
+                        id=f"{self._error_id_prefix}.E004",
+                    )
+                )
+        viewed = set()
+        for field_name, _ in self.fields_orders:
+            if field_name in viewed:
+                errors.append(
+                    Error(
+                        f"Field '{field_name}' is defined more than once. Vector and filter"
+                        " fields must use distinct field names.",
+                        obj=self,
+                        hint="If you need different configurations for the same field,"
+                        " create separate indexes.",
+                        id=f"{self._error_id_prefix}.E005",
+                    )
+                )
+            viewed.add(field_name)
+            field_ = model._meta.get_field(field_name)
+            if isinstance(field_, ArrayField):
+                try:
+                    int(field_.size)
+                except (ValueError, TypeError):
+                    errors.append(
+                        Error(
+                            f"Atlas vector search requires size on {field_name}.",
+                            obj=self,
+                            id=f"{self._error_id_prefix}.E001",
+                        )
+                    )
+                if not isinstance(field_.base_field, FloatField | DecimalField):
+                    errors.append(
+                        Error(
+                            "Base type must be Float or Decimal.",
```

Comment:
"An Atlas vector search index requires the base field of ArrayField Model.field_name to be FloatField or DecimalField but is <field_class>."